dali/dfuplus/dfuplus.cpp

```diff
@@ -1503,112 +1503,137 @@ int CDfuPlusHelper::listhistory()
-                            first = false;
-
-                        const char *propName = attribIter->queryName();
-                        propName++;
-                        csvDump.append(propName);
-                    }
-                    csvDump.append("\n");
-                }
+                // header
+                csvDump.append("ip,name,operation,owner,path,timestamp,workunit\n");
             }
 
-            ForEach(*historyIter)
+            for(unsigned i = 0; i < arrHistory.length(); i++)
             {
-                first = true;
-                Owned<IAttributeIterator> attribIter = historyIter->query().getAttributes(sorted);
-                ForEach(*attribIter)
+                IConstHistory* historyRecord = &arrHistory.item(i);
+                if(historyRecord != nullptr)
                 {
-                    if (!first)
-                        csvDump.append(",");
-                    else
-                        first = false;
-
-                    const char *propVal =  attribIter->queryValue();
-                    csvDump.append(propVal);
+                    csvDump.appendf("%s,", historyRecord->getIP());
+                    csvDump.appendf("%s,", historyRecord->getName());
+                    csvDump.appendf("%s,", historyRecord->getOperation());
+                    csvDump.appendf("%s,", historyRecord->getOwner());
+                    csvDump.appendf("%s,", historyRecord->getPath());
+                    csvDump.appendf("%s,", historyRecord->getTimestamp());
+                    csvDump.appendf("%s\n",  historyRecord->getWorkunit());
                 }
-                csvDump.append("\n");
             }
             info("%s", csvDump.str());
         }
         break;
 
-    case json:
+        case json:
         {
-            StringBuffer out;
-            toJSON(history, out, 3);
-            info("\n%s\n",out.trim().str());
+            StringBuffer jsonDump;
+            const char * level1 = "    ";        // 4 space 'tab'
+            const char * level2 = "        ";
+            const char * level3 = "            ";
+            //                     123456789012
+
+            unsigned index = 1;
+
+            jsonDump.append("{\n");       // Level 0: Start History object
+
+            jsonDump.appendf("%s\"%s\": [", level1, "History");   // Begin History object's array of historical records
+
+            for(unsigned i = 0; i < arrHistory.length(); i++)
+            {
+                if (index++ != 1)
+                    jsonDump.append(",");
+
+                jsonDump.appendf("\n%s{\n", level2);    // Historical record object begin
+
+                bool first = true;
+                IConstHistory* historyRecord = &arrHistory.item(i);
+                if(historyRecord != nullptr)
+                {
+                    // Historical record item object
+                    jsonDump.appendf("%s\"ip\" : \"%s\",\n" ,level3, historyRecord->getIP());
```

Comment:
There are functions to make this easier.
For example:
  appendJSONValue(jsonDump, "ip", historyRecord->getIP());

Hides all of the formatting so it's easier to read, you just need to add padding.